packages/normalizr/src/denormalize.ts

```diff
@@ -21,72 +32,131 @@ const unvisitEntity = (
 
-  return [
-    function unvisit(input: any, schema: any): [any, boolean, boolean] {
-      if (!schema) return [input, true, false];
-
-      if (!schema.denormalize || typeof schema.denormalize !== 'function') {
-        if (typeof schema === 'function') {
-          if (input instanceof schema) return [input, true, false];
-          return [new schema(input), true, false];
-        } else if (typeof schema === 'object') {
-          const method = Array.isArray(schema)
-            ? arrayDenormalize
-            : objectDenormalize;
-          return method(schema, input, unvisit);
-        }
+  function unvisit(
+    input: any,
+    schema: any,
+  ): [denormalized: any, found: boolean, deleted: boolean] {
+    if (!schema) return [input, true, false];
+
+    if (!schema.denormalize || typeof schema.denormalize !== 'function') {
+      if (typeof schema === 'function') {
+        if (input instanceof schema) return [input, true, false];
+        return [new schema(input), true, false];
+      } else if (typeof schema === 'object') {
+        const method = Array.isArray(schema)
+          ? arrayDenormalize
+          : objectDenormalize;
+        return method(schema, input, wrappedUnvisit);
       }
+    }
 
-      // null is considered intentional, thus always 'found' as true
-      if (input === null) {
-        return [input, true, false];
-      }
+    // null is considered intentional, thus always 'found' as true
+    if (input === null) {
+      return [input, true, false];
+    }
 
-      if (isEntity(schema)) {
-        // unvisitEntity just can't handle undefined
-        if (input === undefined) {
-          return [input, false, false];
-        }
-        return unvisitEntity(input, schema, unvisit, getEntity, cache);
+    if (isEntity(schema)) {
+      // unvisitEntity just can't handle undefined
+      if (input === undefined) {
+        return [input, false, false];
       }
+      return unvisitEntity(
+        input,
+        schema,
+        wrappedUnvisit,
+        getEntity,
+        localCache,
+        entityCache,
+      );
+    }
 
-      if (typeof schema.denormalize === 'function') {
-        return schema.denormalize(input, unvisit);
-      }
+    if (typeof schema.denormalize === 'function') {
+      return schema.denormalize(input, wrappedUnvisit);
+    }
 
-      return [input, true, false];
-    },
-    cache,
-  ] as const;
+    return [input, true, false];
+  }
+
+  // make unvisit that tracks entity list here
+  const globalKey: EntityInterface[] = [];
+  const wrappedUnvisit = (input: any, schema: any) => {
+    const ret: [any, boolean, boolean] = unvisit(input, schema);
+    // pass over undefined in key
+    if (ret[0] && schema && isEntity(schema)) globalKey.push(ret[0]);
+    return ret;
+  };
+  wrappedUnvisit.original = unvisit;
+
+  return (
+    input: any,
+    schema: any,
+  ): [denormalized: any, found: boolean, deleted: boolean] => {
+    globalKey.push(input);
```

Comment:
same here?
```suggestion
    input: EntityInterface,
    schema: any,
  ): [denormalized: any, found: boolean, deleted: boolean] => {
    globalKey.push(input);
```

Reply:
`unvisit` will run on any part of input, which can be things that aren't entities

e.g., in the input:
`[{ car: { id: 1, name: 'tesla' }, items: 5 }, { car: { id: 1, name: 'ford' }, items: 2 }]`

each level will run, so we'll see array, then { car, items}, then finally the { id, name} piece. only that last one is an entity.

Reply:
ah, so only the properly typed thing will be pushed into `globalKey`?

Reply:
exactly, that's what the if statement is for